src/socket/encrypter.rs

```diff
@@ -0,0 +1,250 @@
+            send_outgoing: send_out,
+            send_incoming: send_in,
+            _task: Arc::new(task),
+        }
+    }
+
+    pub fn server_new(
+        my_isk: IdentitySecret,
+        remote_ep: Endpoint,
+        rendezvous_fp: Option<Fingerprint>,
+        n2r_skt: N2rSocket,
+        send_incoming_decrypted: Sender<(Bytes, Endpoint)>,
+        client_hs: Handshake,
+        ctx: DaemonContext,
+    ) -> Self {
+        let (send_out, recv_out) = smol::channel::bounded(1000);
+        let (send_in, recv_in) = smol::channel::bounded(1000);
+        let task = smolscale::spawn(enc_task(
+            my_isk,
+            n2r_skt,
+            remote_ep,
+            rendezvous_fp,
+            recv_in,
+            recv_out,
+            send_incoming_decrypted,
+            Some(client_hs),
+            ctx,
+        ));
+        Self {
+            send_outgoing: send_out,
+            send_incoming: send_in,
+            _task: Arc::new(task),
+        }
+    }
+
+    pub async fn send_outgoing(&self, msg: Bytes) -> Result<(), SocketSendError> {
+        self.send_outgoing
+            .send(msg)
+            .await
+            .map_err(|_| SocketSendError::HavenSendError)
+    }
+
+    pub async fn send_incoming(&self, msg: HavenMsg) -> Result<(), SocketRecvError> {
+        self.send_incoming
+            .send(msg)
+            .await
+            .map_err(|_| SocketRecvError::HavenRecvError)
+    }
+}
+
+async fn enc_task(
+    my_isk: IdentitySecret,
+    n2r_skt: N2rSocket,
+    remote_ep: Endpoint,
+    rendezvous_fp: Option<Fingerprint>,
+    recv_incoming: Receiver<HavenMsg>,
+    recv_outgoing: Receiver<Bytes>,
+    send_incoming_decrypted: Sender<(Bytes, Endpoint)>,
+    client_hs: Option<Handshake>,
+    ctx: DaemonContext,
+) -> anyhow::Result<()> {
+    // get rendezvous ep
+    let rendezvous_ep = match rendezvous_fp {
+        Some(rob) => {
+            // We're the server
+            Endpoint::new(rob, HAVEN_FORWARD_DOCK)
+        }
+        None => {
+            // We're the client: look up Rob's addr in rendezvous dht
+            log::trace!(
+                "alice is about to send an earendil packet! looking up {} in the DHT",
+                remote_ep.fingerprint
+            );
+            let bob_locator = ctx
+                .dht_get(remote_ep.fingerprint)
+                .await
+                .map_err(|_| SocketSendError::DhtError)?
+                .context("Could not get rendezvous for {endpoint}")
+                .map_err(|_| SocketSendError::HavenSendError)?;
+            log::trace!("found rob in the DHT");
+            Endpoint::new(bob_locator.rendezvous_point, HAVEN_FORWARD_DOCK)
+        }
+    };
+    // complete handshake to get the shared secret
+    let my_osk = OnionSecret::generate();
+    let my_hs = Handshake::new(&my_isk, &my_osk);
+    let shared_sec = match client_hs {
+        Some(hs) => {
+            hs.id_pk.verify(hs.to_sign().as_bytes(), &hs.sig)?; // verify sig
```

Comment:
We also have to pass in the purported source fingerprint, and verify that it corresponds to the identity declared here.

Otherwise, people can trivially fake their source fingerprints. N2R does not provide source authentication, but haven encryption can do so "for free", so we should do it.